src/components/AddressBadge.tsx

```diff
@@ -17,42 +17,77 @@ along with the library. If not, see <http://www.gnu.org/licenses/>.
 */
 
 import { ComponentPropsWithoutRef } from 'react'
-import styled, { css } from 'styled-components'
-import tinycolor from 'tinycolor2'
+import styled, { useTheme } from 'styled-components'
 
+import { Address } from '../contexts/addresses'
+import { AddressSettings, isAddress, isAddressSettings } from '../utils/addresses'
 import Badge from './Badge'
 
 type AddressBadgeProps = ComponentPropsWithoutRef<typeof Badge> & {
-  addressName: string
-  opaqueBg?: boolean
+  address: Address | AddressSettings
+  truncate?: boolean
 }
 
-const AddressBadge = ({ addressName, className, ...props }: AddressBadgeProps) => (
-  <Badge className={className} rounded {...props}>
-    {addressName}
-  </Badge>
-)
-
-export default styled(AddressBadge)`
-  ${({ color, opaqueBg, theme }) => {
-    const usedColor = color || theme.font.secondary
-    const isBright = tinycolor(usedColor).getBrightness() > 160
-
-    return css`
-      color: ${opaqueBg
-        ? isBright
-          ? theme.font.primary
-          : theme.font.contrastPrimary
-        : isBright
-        ? tinycolor(usedColor).darken(5).toString()
-        : usedColor};
-      background-color: ${({ theme }) =>
-        opaqueBg
-          ? tinycolor(usedColor).setAlpha(0.8).toString()
-          : theme.name === 'dark'
-          ? tinycolor(usedColor).setAlpha(0.08).toString()
-          : tinycolor(usedColor).setAlpha(0.1).toString()};
-      padding: 6px 10px;
+const AddressBadge = ({ address, className, ...props }: AddressBadgeProps) => {
+  const theme = useTheme()
+
+  let data
+
+  if (isAddress(address)) {
+    data = {
+      color: address.settings.color,
+      isMain: address.settings.isMain,
+      label: address.getLabelName(false)
+    }
+  } else if (isAddressSettings(address)) {
+    data = address
+  } else {
+    data = {
+      color: theme.font.primary,
+      isMain: false,
+      label: ''
+    }
+  }
+
+  return (
+    <Container className={className} {...props}>
+      <Icon isMain={data.isMain} color={data.color} /> <Label {...props}>{data.label}</Label>
+    </Container>
+  )
+}
+
+export default AddressBadge
+
+const Label = styled.span<AddressBadgeProps>`
+  ${({ truncate }) => truncate && 'overflow: hidden; text-overflow: ellipsis;'}
+`
+
+const Icon = styled.span<{ isMain: boolean; color?: string }>`
+  ${({ color, isMain, theme }) => `
+    &::before {
+      width: 1rem;
+      display: block;
+      text-align: center;
+      margin-right: 0.2rem;
+      line-height: 1rem;
+      font-size: ${isMain ? '1.4em' : '1em'};
+      vertical-align: middle;
+      content: '${isMain ? '★' : '●'}';
+      border-radius: 100%;
+      color: ${color ?? theme.font.primary};
+    }
+  `}
+`
+
+const Container = styled.div<AddressBadgeProps>`
+  padding: 6px 0 6px 0;
+  display: flex;
+
+  ${({ truncate }) =>
+    truncate &&
     `
-  }}
+      white-space: nowrap;
+      overflow: hidden;
+      text-overflow: ellipsis;
+  `}
```

Comment:
Can be rewritten as:
```suggestion
export default styled(AddressBadge)`
  padding: 6px 0 6px 0;
  display: flex;
  
  ${({ truncate }) =>
    truncate &&
    `
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
  `}
```